src/Workspaces/Core/Portable/Storage/CloudCache/ProjectCacheContainerKey.cs

```diff
@@ -0,0 +1,99 @@
+        /// </summary>
+        public readonly CloudCacheContainerKey? ContainerKey;
+
+        /// <summary>
+        /// Cache from document green nodes to the container keys we've computed for it. We can avoid computing these
+        /// container keys when called repeatedly for the same documents.
+        /// </summary>
+        private readonly ConditionalWeakTable<TextDocumentState, StrongBox<CloudCacheContainerKey?>> _documentToContainerKey = new();
+        private readonly ConditionalWeakTable<TextDocumentState, StrongBox<CloudCacheContainerKey?>>.CreateValueCallback _documentToContainerKeyCallback;
+
+        public ProjectCacheContainerKey(string relativePathBase, ProjectKey projectKey)
+        {
+            ContainerKey = CreateProjectContainerKey(relativePathBase, projectKey);
+
+            _documentToContainerKeyCallback = ds => new(CreateDocumentContainerKey(
+                relativePathBase, DocumentKey.ToDocumentKey(projectKey, ds)));
+        }
+
+        public CloudCacheContainerKey? GetValue(TextDocumentState state)
+            => _documentToContainerKey.GetValue(state, _documentToContainerKeyCallback).Value;
+
+        public static CloudCacheContainerKey? CreateProjectContainerKey(
+            string relativePathBase, ProjectKey projectKey)
+        {
+            // Creates a container key for this project.  THe container key is a mix of the project's name, relative
+            // file path (to the solution), and optional parse options.
+
+            // If we don't have a valid solution path, we can't store anything.
+            if (string.IsNullOrEmpty(relativePathBase))
+                return null;
+
+            // We have to have a file path for this project
+            if (string.IsNullOrEmpty(projectKey.FilePath))
+                return null;
+
+            // The file path has to be relative to the base path the DB is associated with (either the solution-path or
+            // repo-path).
+            var relativePath = PathUtilities.GetRelativePath(relativePathBase, projectKey.FilePath!);
+            if (relativePath == projectKey.FilePath)
+                return null;
+
+            var dimensions = EmptyDimensions
+                .Add($"{nameof(ProjectKey)}.{nameof(ProjectKey.Name)}", projectKey.Name)
+                .Add($"{nameof(ProjectKey)}.{nameof(ProjectKey.FilePath)}", relativePath)
+                .Add($"{nameof(ProjectKey)}.{nameof(ProjectKey.ParseOptionsChecksum)}", projectKey.ParseOptionsChecksum.ToString());
+
+            return new CloudCacheContainerKey("Roslyn.Project", dimensions);
+        }
+
+        public static CloudCacheContainerKey? CreateDocumentContainerKey(
+            string relativePathBase,
+            DocumentKey documentKey)
+        {
+            // See if we can get a project key for this info.  If not, we def can't get a doc key.
+            var projectContainerKey = CreateProjectContainerKey(relativePathBase, documentKey.Project);
```

Comment:
so tehre are two types of 'keys' being discussed here.  One is that the IPersistenceService itself allows you to read/write data using lightweight keys, instead of passing in a full Document/Project instance.  This is used for cases where we want to store data assocaited with a doc, but we don't care if the doc is from teh Release config or Debug config, and so we manually generate the keys to be broader that what the default generally is.

THe second type of 'key' are the keys used by the platform's caching system that we're sitting on top of.  This code translates between the two, trying to ensure we do minimal allocations over tiem.  i.e. if many things are commonly reading/writing from a particular roslyn doc key, we can alloc only once when creating the platform key.